src/coreclr/vm/clsload.cpp

```diff
@@ -91,33 +118,42 @@ PTR_Module ClassLoader::ComputeLoaderModuleWorker(
 
     Module *pLoaderModule = NULL;
 
-    if (pDefinitionModule)
-    {
-        if (pDefinitionModule->IsCollectible())
-            goto ComputeCollectibleLoaderModule;
-        pLoaderModule = pDefinitionModule;
-    }
+    if (pDefinitionModule != NULL && pDefinitionModule->IsCollectible())
+        goto ComputeCollectibleLoaderModule;
 
     for (DWORD i = 0; i < classInst.GetNumArgs(); i++)
     {
         TypeHandle classArg = classInst[i];
+
+        // System.__Canon does not contribute to logical loader module
+        if (classArg == TypeHandle(g_pCanonMethodTableClass))
+            continue;
         Module* pModule = classArg.GetLoaderModule();
         if (pModule->IsCollectible())
             goto ComputeCollectibleLoaderModule;
-        if (pLoaderModule == NULL)
+        if ((pLoaderModule == NULL) && (pModule != pDefinitionModule))
             pLoaderModule = pModule;
     }
 
     for (DWORD i = 0; i < methodInst.GetNumArgs(); i++)
     {
         TypeHandle methodArg = methodInst[i];
+
+        // System.__Canon does not contribute to logical loader module
+        if (methodArg == TypeHandle(g_pCanonMethodTableClass))
+            continue;
         Module *pModule = methodArg.GetLoaderModule();
         if (pModule->IsCollectible())
             goto ComputeCollectibleLoaderModule;
-        if (pLoaderModule == NULL)
+        if ((pLoaderModule == NULL) && (pModule != pDefinitionModule))
             pLoaderModule = pModule;
     }
 
+    if ((pLoaderModule == NULL) && pDefinitionModule)
```

Comment:
Nit - the check for <code>pDefinitionModule</code> isn't really necessary here, I guess it's just a matter of your aesthetic feeling ;-).